### ISSUE TITLE
After upgrade to 9.x - aws_lb_target_group_attachment does not have an attribute named "target_id"

### ISSUE THREAD

**Issue by falc410** [terraform-aws-modules/terraform-aws-alb#321]
## Description

I copied your example https://github.com/terraform-aws-modules/terraform-aws-alb/blob/master/README.md#http-to-https-redirect

and replaced my previous configuration. However when I try to run this, I get an error

```
Error: Unsupported attribute
 
   on .terraform/modules/routing.alb/main.tf line 511, in resource "aws_lb_target_group_attachment" "this":
  511:   target_id         = each.value.target_id
     │ each.value is object with 4 attributes
 
 This object does not have an attribute named "target_id".

```
I see that in the Migration / Upgrade Guide to 9.x you added `target_id = aws_instance.this.id` however, it is not included in the example of the main README and I didn't use this before. Also, I am not sure where aws_instance.this is defined in the example. Since I'm using Fargate Containers my target_type = "ip" and not "instance" - might this be the problem?

```
module "alb" {
  source = "terraform-aws-modules/alb/aws"
  create_security_group = false # see https://github.com/terraform-aws-modules/terraform-aws-alb/pull/273
  enable_xff_client_port = false # new feature, enabled by default but AWS defaults to false 
  name = "r1-alb-${var.environment}"

  load_balancer_type = "application"

  vpc_id          = var.vpc.vpc_id
  security_groups = var.sg
  subnets         = var.vpc.public_subnets

  
  access_logs = {
    bucket = aws_s3_bucket.logs.id
  }

  listeners = {
    http-https-redirect = {
      port     = 80
      protocol = "HTTP"
      redirect = {
        port        = "443"
        protocol    = "HTTPS"
        status_code = "HTTP_301"
      }
    }
    https = {
      port            = 443
      protocol        = "HTTPS"
      certificate_arn = module.acm_primary.acm_certificate_arn

      forward = {
        target_group_key = "instance"
      }
    }
  }
  target_groups = {
    instance = {
      name_prefix      = "r1"
      protocol         = "HTTP"
      port             = 80
      target_type      = "ip"
    }
  }
} # end module alb
```

- [ x] ✋ I have searched the open/closed issues and my issue is listed but unresolved, see: https://github.com/terraform-aws-modules/terraform-aws-alb/issues/317

## Versions

- Module version [Required]: 9.1.0
- Terraform version: 1.5.7
- Provider version(s): aws v5.24.0

## Edit

It seems I can prevent this error message by adding `create_attachment = false` to the target_group Definition, though I don't really understand the implications of this. At least it seems to work

**Comment by trevorrea** [comment]
The implication of create_attachment = false is to make the target_id attribute optional. This is covered in the docs at https://github.com/terraform-aws-modules/terraform-aws-alb/blob/master/docs/patterns.md#target-group-without-attachment and in the code at https://github.com/terraform-aws-modules/terraform-aws-alb/blob/master/main.tf#L508 

As you can see if create_attachment is not set to false it will try to create the target group attachment and fail as target_id is not set.

ECS manages target group attachments itself so as at https://github.com/terraform-aws-modules/terraform-aws-alb/blob/master/docs/patterns.md#target-group-without-attachment create_attachment needs to be false.

It's not a bug and is pretty well documented.

**Comment by falc410** [comment]
> It's not a bug and is pretty well documented.

Well, I disagree on the latter part. The official example in the README doesn't work (I copied the code from there) and it is not mentioned in the 9.x Upgrade Guide as another breaking change. I didn't notice the usage patterns doc though - I guess it could have been found there. But that's not what I would call "pretty well documented".  I think this could be improved.

**Comment by bryantbiggs** [comment]
I am open to feedback to improve the docs, and I'd love to know what part is incorrect on the README example as well

**Comment by falc410** [comment]
I mentioned it in the first post, this part here: https://github.com/terraform-aws-modules/terraform-aws-alb#http-to-https-redirect
I basically copied your example which creates two listeners (redirect http to https and https listener) and one example target_group but it doesn't include the `create_attachment = false` 
Seeing as I am not the only one who ran into this problem, I would suggest to update the example.
In the docs/patterns.md there is a functioning example, even with comments: https://github.com/terraform-aws-modules/terraform-aws-alb/blob/master/docs/patterns.md#target-group-without-attachment

**Comment by bryantbiggs** [comment]
The README example is not deployable - its meant to give users a reference when first viewing the module, and its not tied to any deployment means. Unfortunately, this means that sometimes there are small typos/errors (since it cannot be validated)

The https://github.com/terraform-aws-modules/terraform-aws-alb/tree/master/examples are deployable and this is what we use to both test/validate the module in addition to demonstrate the breadth of capabilities. 

Lastly, a bit of empathy would go a long way - this body of work is provided free of charge so I don't think that tone or demeanor is warranted

**Comment by falc410** [comment]
> Lastly, a bit of empathy would go a long way - this body of work is provided free of charge so I don't think that tone or demeanor is warranted

to quote your words: I'd love to know what part you are referring to. 

All I did say, is that (as a user), I expected the example in the README (which is the first thing you see when you go to this repository) to be correct.  You wanted a suggestion and I gave you a suggestion, namely to update the example code. If the example code of the README is not deployable / useable, I suggest mentioning this at top of the README. To me, it was not clear.
